### ISSUE TITLE
cpyext: PyObject_GetBuffer() returns NUL-filled buffer (instead of data) when called intensively

### ISSUE THREAD

**Issue by gitlab-importer** [pypy/pypy#2752]
In Heptapod by bitbucket_importer on Feb 12, 2018, 20:06

*Created originally on Bitbucket by [mgorny (Michał Górny)](https://bitbucket.org/[7.3.7](https://github.com/gitlab-importer/pypy/milestone/7)B086a5ef0-f571-46e6-b7f2-84e5ccc49442%7D/)*

I've noticed that my program that hashes intensively using [pyblake2](https://github.com/dchest/pyblake2) extension starts giving wrong results at some point. I haven't been able to establish what is the exact cause but I've been able to create a [test case](https://github.com/mgorny/pypy-blake2-testcase) that easily reproduces the problem.

The exact code is:

```
#!python

import io
import pyblake2

bufsize = 4096
sum = 'ccefcd101b08863339602f7fdf2edd1d77ef05a970c36dbd7a560d33f957f81b15cfcac10114f8fca0d7c318b6aaa294220e3fcf4f88e6e3bd7840f121ff3b65'

def sub(i):
    cs = pyblake2.blake2b()
    with io.open('test.txt', 'rb') as f:
        for block in iter(lambda: f.read(bufsize), b''):
            cs.update(block)
    assert cs.hexdigest() == sum, i

for x in map(sub, range(10000)):
    pass
```

With this test case, PyPy reliably fails (generates incorrect checksum) at iteration 94.

Few observations based on testing:

1. The issue affects PyPy2 only. PyPy3 and CPython work fine.
2. I can reproduce a similar problem with pyblake2, pysha3 but not e.g. pycryptodome (which is also C extension) or builtin hash functions.
3. Some random changes to code (e.g. replacing io.open() with open()) cause the failing iteration no to change.
4. If instead of the loop, I do a single `f.read()`, I wasn't able to get it to fail (even with increased iteration count).
5. If I do `f.read()` without argument in a loop, it fails at iteration 852.
6. Changing `bufsize` and iteration count also changes the result, with no clear correlation. E.g. with bufsize of 506000 and 10000 iterations, it doesn't fail. With 100000 iterations, it fails at iteration 1488...

In other words, I really have no clue what might be happening here. I'm attaching the test script and file for completeness.

 *Attachments:* [test.py](https://pypy-attachments.s3.amazonaws.com/736937129/a1ee37d9385e39252183625593a769d89392f5fdf4354e6523b3c9734ffc75b3/test.py)[testmod.c](https://pypy-attachments.s3.amazonaws.com/736937129/20cc7caac1a4877cb02d317febf5e47bd5bb66702da5b6d431967e1f8f4108f9/testmod.c)
